ggml-cuda.cu

```diff
@@ -6258,6 +6265,41 @@ static void ggml_cuda_op_flatten(const ggml_tensor * src0, const ggml_tensor * s
     }
 }
 
+void ggml_cuda_set_peer_access(const int n_tokens) {
+    static int last_n_tokens = INT_MAX;
+
+    if (n_tokens == last_n_tokens) {
+        return;
+    }
+
+#ifdef NDEBUG
+    for (int id = 0; id < g_device_count; ++id) {
+        CUDA_CHECK(cudaSetDevice(id));
+
+        for (int id_other = 0; id_other < g_device_count; ++id_other) {
+            if (id == id_other) {
+                continue;
+            }
+            if (id != g_main_device && id_other != g_main_device) {
+                continue;
+            }
+
+            int canAccessPeer;
+            CUDA_CHECK(cudaDeviceCanAccessPeer(&canAccessPeer, id, id_other));
+            if (canAccessPeer) {
+                if (n_tokens <= GGML_CUDA_PEER_MAX_BATCH_SIZE && last_n_tokens > GGML_CUDA_PEER_MAX_BATCH_SIZE) {
+                    CUDA_CHECK(cudaDeviceEnablePeerAccess(id_other, 0));
+                } else if (n_tokens > GGML_CUDA_PEER_MAX_BATCH_SIZE && last_n_tokens <= GGML_CUDA_PEER_MAX_BATCH_SIZE) {
```

Comment:
What I meant here is that the processing could be avoided entirely if peer access doesn't need to be changed. There is already a check for `n_tokens == last_n_tokens`, so it could be extended to something like this:
```C++
    if ((n_tokens <= GGML_CUDA_PEER_MAX_BATCH_SIZE && last_n_tokens <= GGML_CUDA_PEER_MAX_BATCH_SIZE) || (n_tokens > GGML_CUDA_PEER_MAX_BATCH_SIZE && last_n_tokens > GGML_CUDA_PEER_MAX_BATCH_SIZE)) {
        return;
    }
```

Or, if `last_n_tokens` was replaced by a boolean that indicates whether peer access is currently enabled, the logic could be even simpler:
```C++
static bool last_enable = false;
bool enable = n_tokens <= GGML_CUDA_PEER_MAX_BATCH_SIZE;

if (last_enable != enable) {
    return;
}
```

Ultimately, the goal is to simplify the code more than anything else, duplicating the loops is the opposite of what I was suggesting here.

Reply:
Alright, I was thinking your goal was to reduce the number of CUDA calls because as recent experience has shown they can have bad performance on some OSs.